lib/srv/reexec.go

```diff
@@ -764,39 +660,203 @@ func runForward(handler forwardHandler) (errw io.Writer, code int, err error) {
 	for {
 		buf := make([]byte, 1024)
 		fbuf := make([]*os.File, 1)
-		n, fn, err := uds.ReadWithFDs(conn, buf, fbuf)
+		n, fn, err := uds.ReadWithFDs(parentConn, buf, fbuf)
 		if err != nil {
 			if utils.IsOKNetworkError(err) {
-				return errorWriter, teleport.RemoteCommandSuccess, nil
+				// parent connection closed, process should exit.
+				return errorWriter, teleport.RemoteCommandFailure, nil
 			}
-			return errorWriter, teleport.RemoteCommandFailure, trace.Wrap(err)
+			slog.With("error", err).ErrorContext(ctx, "Error reading networking request from parent.")
+			continue
 		}
-		addr := string(buf[:n])
+
 		if fn == 0 {
-			log.Errorf("Parent did not send a file descriptor for address %q.", addr)
+			slog.ErrorContext(ctx, "Networking request requires a control file.")
 			continue
 		}
 
-		go func() {
-			if err := handler(ctx, addr, fbuf[0]); err != nil {
-				log.WithError(err).Errorf("Error handling forwarding request for address %q.", addr)
+		controlConn, err := uds.FromFile(fbuf[0])
+		_ = fbuf[0].Close()
+		if err != nil {
+			slog.ErrorContext(ctx, "Failed to get a connection from control file.")
+			continue
+		}
+
+		var req networking.Request
+		if err := json.Unmarshal(buf[:n], &req); err != nil {
+			slog.With("error", err).ErrorContext(ctx, "Error parsing networking request.")
+			continue
+		}
+
+		// Some PAM modules (e.g. pam_namespace) do not behave properly in multithreaded contexts.
+		// Therefore we favor handling requests on the main PAM thread for requests that are expected
+		// to be impacted (unix socket listeners).
+		switch req.Operation {
+		case networking.NetworkingOperationDial, networking.NetworkingOperationListen:
+			switch req.Network {
+			case "tcp":
+				// There are currently no known issues with tcp listen/dial in a multithreaded PAM context.
+				go handleNetworkingRequest(ctx, controlConn, req)
```

Comment:
Until someone comes along and really wants to use a random pam module that does network namespacing 😬